Add new C++ and Rust templates for GUI applications written in Slint

tronical opened the pull request.

tronical committed:
cppSlint: Add a new C++ template for a GUI application written in Slint

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical commented:
I've tried this on a imx8 on a dahlia board and it works for me. The docker file is a little complex, and it also takes a fair while to build. The upside though is that since the slint building part is the same for the SDK, the deploy image for the SDK, as well as the production, once it's built those layers are reused.

tronical commented:
Regarding the logo: Do you know by chance if SVGs are supported, too, by your extension? That would look better than the PNG file for the Slint logo.

tronical committed:
Minor tweak to the CppTools config

Slint requires C++20

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: Improve the logo

Use the SVG version, which also includes the product name as text.

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: Remove unnecessary dependencies

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: simplify update.json

We don't need to deal with qmake here :-)

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: Add expected debug output message

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: Fix embedding of external images into the binary

This way it just works out of the box

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: Default to using the Skia renderer with Slint

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: Default to Slint running as fullscreen app

That works best on devices and it seems that's what the other templates
do as well.

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: Make the Slint template app scale better in fullscreen

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: Replace clangd with ms cpptools

Both installed gives constant messages about them conflicting.

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: Fix source file mapping when debugging

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: Fix debug source mapping for Slint compiler generated code

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
rustSlint: Add support for building GUI applications in Rust with Slint

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
cppSlint: Fix file name and line for debugger tests

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
rustSlint: Remove local AMD64 task

This isn't supported in the docker base images. But rust-analyzer can
generate the correct cargo tasks, so leave it out of this template.

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical committed:
rustSlint: Provide local run and debug launch configs

The auto-generation with rust-analyzer is good, but it requires manual
merging into launch.json - let's do that for the user instead.

microhobby commented:
@tronical sorry I only see this now, but all the new development should be merged into the `dev` branch. Could you please use the `dev` branch as base for your development?

Tip: I think that will be easier to you get your commits and cherry-pick to a fresh rebase from dev than try to rebase `dev` into `bookworm`

tronical commented:
Sure, I'll rebase onto dev.

tronical commented:
Done. There were no conflicts. Did dev diverge from bookworm in a way that would require making changes to the docker files, etc?

microhobby commented:
@tronical oh cool, not differ in content, only in commit hashes, so I thought that this will be an issue. But good to know. Thanks!

microhobby commented on cppSlint/.conf/deps.json:
```diff
@@ -0,0 +1,12 @@
+        "gdb",
+        "gdb-multiarch",
+        "g++",
+        "netcat-traditional"
```
This package is only needed from Qt because we do some port redirect to use the Qt Creator as debugger. Are you using it for something in the Slint template?

tronical replied:
Indeed, that's a leftover from copy & paste. Will remove - Thanks :)

microhobby commented on cppSlint/.vscode/c_cpp_properties.json:
```diff
@@ -0,0 +1,17 @@
+            "name": "Linux",
+            "includePath": [
+                "${workspaceFolder}/**",
+                "/usr/include/**"
```
There is some way to download and configure here the Slint headers, like the `appwindow.h` for example?

microhobby replied:
The cpp linter works locally, and we should have it locally to set in the `includePath`.
Maybe to add a task entry to be executed in the project open to download the headers...

tronical replied:
Hm, `appwindow.h` is a generated header file (from `appwindow.slint`), so I'm not sure where that could be downloaded from/to. The Slint C++ API headers are inside the docker image and installed into /usr/include.

How do you handle this with other C/C++ libraries?

microhobby replied:
We usually instruct users to install the reference dev package `yourlib-dev`, or download the headers to have a local reference. As I said, the linter runs locally, I mean outside the SDK container.

But from my tests the `build-debug-amd64-local`, already get the `build/x86_64/_deps/slint-src/api/cpp/include` folder that has the API headers. So, LGTM

microhobby commented on cppSlint/Dockerfile.sdk:
```diff
@@ -0,0 +1,106 @@
+    --mount=type=cache,target=/var/cache/apt \
+    CROSS_TOOLCHAIN_ARCH=$(cat /cross-toolchain-arch.txt) && \
+    apt-get update && \
+    apt-get install --assume-yes libfontconfig1-dev:$CROSS_TOOLCHAIN_ARCH libxcb1-dev:$CROSS_TOOLCHAIN_ARCH libxcb-render0-dev:$CROSS_TOOLCHAIN_ARCH libxcb-shape0-dev:$CROSS_TOOLCHAIN_ARCH libxcb-xfixes0-dev:$CROSS_TOOLCHAIN_ARCH libxkbcommon-dev:$CROSS_TOOLCHAIN_ARCH python3 clang libstdc++-10-dev:$CROSS_TOOLCHAIN_ARCH && \
```
`libstdc++-10-dev` has no candidate for Debian `bookworm`. Is possible to use the `libstdc++-11-dev`? https://packages.debian.org/search?suite=bookworm&section=all&arch=any&searchon=names&keywords=libstdc%2B%2B-

tronical replied:
Yes, that's the correct choice, I'll fix that. When I started, the 3-bookwork tag only had libstdc++10-dev, but it seems that meanwhile you moved the tag to a newer Debian snapshot where 10 was replaced with 11.

microhobby commented on cppSlint/Dockerfile.sdk:
```diff
@@ -0,0 +1,106 @@
+
+USER torizon
+
+WORKDIR /app
```
yeah, this take some time to build. Maybe should be good to you guys provide some pre-built Docker images to use as base, like `slint/slint-ui` for `arm64v8` and `arm32v7`. This should greatly improve build speed.

microhobby replied:
The comment above is only a suggestion, this will not block the merge.

tronical replied:
Yeah, it's tempting to do that. But it's also a bit of work to maintain them in parallel your images. When you update your 3-bookworm tag, we need to update our images, too. Another option is to entirely deviate from Torizon, but then we might run into issues where our container creates run-time library dependencies that aren't present in the torizon deployment containers.

microhobby commented on assets/img/slint-logo-square-light.svg:
For maintaining the consistency, I would like to propose to use the follow images:
![slintrust](https://user-images.githubusercontent.com/2633321/235315938-b0e8d848-2c4b-4668-b5af-3f30b2cb56af.PNG)
![slintcpp](https://user-images.githubusercontent.com/2633321/235315939-aff97426-23c3-4b81-8aa4-81c35e90e2e4.PNG)

Let me know if this works for you.

tronical replied:
Excellent idea! Let me see if I can integrate those tag images into the .svg, otherwise we can use your png files :)

tronical replied:
Alright, I can't see a trivial way - so I pushed a commit with your png versions. Thanks again :)

microhobby commented:
@tronical for now I only checked the C++ one, next week we will check the Rust template.

microhobby commented on cppSlint/.vscode/settings.json:
```diff
@@ -0,0 +1,19 @@
+  "torizon_run_as": "torizon",
+  "torizon_app_root": "/home/torizon",
+  "tcb.packageName": "__change__",
+  "tcb.version": "early-access"
```
Did you do not want to add here the:

```json
  "slint.preview.providedByEditor": true
```

I was testing here and this integrates very well instead of the default preview that uses a fullscreen window

tronical replied:
I'd prefer to stay with the defaults here. The side-by-side works, but in terms of screen real estate that remains left for the code editor, I prefer the separate preview window.

microhobby replied:
Is there a way to not show the window in full screen, then? My concern is that it makes users, that uses single screen, confused.

microhobby replied:
ok, I tried again and it's showing a normal window, resolving also this one

tronical committed:
cppSlint: Fix docker image building

The 3-bookworm tag no longer includes libstdc++-10 but has moved on top
-11. Use that, just like with the rustSlint template.

tronical committed:
cppSlint: Remove unnecessary package dependency

tronical committed:
Replace Slint SVG logo with C++/Rust tagged versions

microhobby commented on rustSlint/src/main.rs:
```diff
@@ -0,0 +1,15 @@
+        ui.set_counter(ui.get_counter() + 1);
+    });
+
+    println!("Hello, Toradex!");
```
I think here should be `Hello Torizon!`

tronical replied:
Ah yes, well spotted!

microhobby commented:
@tronical thanks a lot for the changes, I check also the rust one and apart from the small detail on the `Hello Torizon!` string, LGTM. As soon as you change this detail we will merge it, thank you.

tronical committed:
rustSlint: Fix expected debug message

Signed-off-by: Simon Hausmann <simon.hausmann@slint-ui.com>

tronical commented:
Thanks for the review! I pushed the fix.
